pyteal/stack_frame.py

```diff
@@ -0,0 +1,701 @@
+            rel_paths=other.rel_paths,
+            parent=self,
+        )
+        ptf._status = status
+
+        return ptf
+
+    def location(self) -> str:
+        return f"{self.file()}:{self.lineno()}" if self.frame_info else ""
+
+    def file(self) -> str:
+        if not self.frame_info:
+            return ""
+
+        path = self.frame_info.filename
+        return os.path.relpath(path) if self.rel_paths else path
+
+    def root(self) -> str:
+        if not self.frame_info:
+            return ""
+
+        path = self.frame_info.filename
+        return path[: -len(self.file())]
+
+    def code_qualname(self) -> str:
+        return (
+            Source.executing(self.frame_info.frame).code_qualname()
+            if self.frame_info
+            else ""
+        )
+
+    def lineno(self) -> int | None:
+        naive_lineno = self.frame_info.lineno if self.frame_info else None
+        if naive_lineno is None:
+            return naive_lineno
+
+        hybrid_line, offset = self._hybrid_w_offset()
+        if hybrid_line.startswith(self.raw_code()):
+            return naive_lineno
+
+        return naive_lineno + offset
+
+    def column(self) -> int:
+        """Provide accurate column info when available. Or 0 when not."""
+        return self.node_col_offset() or 0
+
+    def compiler_generated_reason(self) -> str | None:
+        """
+        None indicates either "unkown" or "not compiler generated".
+        To distinguish between these two usages, call `compiler_generated()` first.
+        """
+        if not self.compiler_generated():
+            return None
+
+        if self.creator._compiler_gen:
+            return PT_GENERATED.FLAGGED_BY_DEV
+
+        for k, v in _PT_GEN.items():
+            if k in self.raw_code():
+                return v
+
+        raise AssertionError(
+            "This should never happen as the call to self.compiler_generated() should have prevented this case."
+        )
+
+    # VARIATIONS ON THE THEME OF SOURCE CODE:
+
+    def raw_code(self) -> str:
+        if self._raw_code is None:
+            self._raw_code = (
+                ("".join(self.frame_info.code_context)).strip()
+                if self.frame_info and self.frame_info.code_context
+                else ""
+            )
+
+        return self._raw_code
+
+    def hybrid_unparsed(self) -> str:
+        """
+        Attempt to unparse the node and return the most apropos line
+        """
+        return self._hybrid_w_offset()[0]
+
+    def _hybrid_w_offset(self) -> tuple[str, int]:
+        """
+        Attempt to unparse the node and return the most apropos line,
+        together with its offset in comparison with its raw code.
+        """
+        code = self.code()
+        node = self.node
+        pt_chunk = self.node_source()
+        return self._hybrid_impl(code, node, pt_chunk)
+
+    @classmethod
+    def _hybrid_impl(
+        cls, code: str, node: Optional[AST], pt_chunk: str
```

Comment:
maybe a line about what you're doing here

Reply:
cf [f615c01](https://github.com/algorand/pyteal/pull/650/commits/f615c019de85f2fbee2b783591b030c20139a160)